src/cargo/core/registry.rs

```diff
@@ -718,3 +743,114 @@ fn lock(
         dep
     })
 }
+
+/// This is a helper for selecting the summary, or generating a helpful error message.
+fn summary_for_patch(
+    orig_patch: &Dependency,
+    locked: &Option<(Dependency, PackageId)>,
+    mut summaries: Vec<Summary>,
+    source: &mut dyn Source,
+) -> CargoResult<(Summary, Option<PackageId>)> {
+    if summaries.len() == 1 {
+        return Ok((summaries.pop().unwrap(), None));
+    }
+    let best_summary = |summaries: &mut Vec<Summary>| -> Summary {
+        // TODO: This could maybe honor -Zminimal-versions?
+        summaries.sort_by(|a, b| a.version().cmp(b.version()));
+        summaries.pop().unwrap()
+    };
+    if summaries.len() > 1 {
```

Comment:
This makes me a bit hesitant here trying to handle multiple candidates being returned. When `[patch]` was originally implemented we didn't have custom registries yet, so the only real usage of `[patch]` was pointing to `git` and `path`, neither of which can possibly have multiple versions of a package. With custom registries, though, that's now possible!

My main worry is that the candidates here don't participate in the normal version resolution of everything else. The biggest version may not always be the right one to pick (with other constraints like `links` and such).

Could this continue to restrict to only one package? I think it's fine to allow more candidates, but I think what should be done is to add all the candidates to "these can be searched through with version resolution". Version resolution would then do its normal thing to pick which is the best.

Reply:
I'm not sure if I'm understanding correctly.  If you're saying that this function should return multiple summaries, the problem is that any of the entries that aren't selected during version resolution end up in the unused patch list which triggers a warning.

I guess it could group related summaries together, so that the "unused patch" logic could only mark them unused if none in the group were used.  However, that would require changes in lots of places (which I'd rather not do right now).

Unless you meant something else, should I just switch it back to an error (maybe with a suggestion to use an `=` version requirement)?

Reply:
Oh sorry, let me clarify. Primarily this is a change in behavior from today which I'm worried about, so I would prefer to switch it back to make multiple candidates an error.

Other than that I was thinking that we probably *could* support this, but not by picking a version here and instead letting version resolution later pick a version. The theoretical idea behind `[patch]` was "take this crate and assume it comes from that source". That same idea works for "take this set of crates and assume they all come from that source", but the code here just isn't structured to do that. Ideally what we should do here is that when there's an unlocked (the only way to return multiple versions) dependency we make all matching candidates available for version resolution from the resolver, and the resolver picks one we'd then lock. If it ends up being entirely unused then we could pick the biggest one to lock and it wouldn't be an issue.

Does that make sense?